### PR TITLE
pkg/generator: Generate all manifests needed for Prometheus integration

### PR DESCRIPTION
**Pull request by etiennecoutaud**
Refer to #222 

Generate kubernetes manifest files to integrate operator with Prometheus

### CODE REVIEW

**Comment by etiennecoutaud** [comment]
cc @spahl

File: pkg/generator/deploy_tmpl.go [hunk 1]
```diff
@@ -33,19 +33,23 @@ const operatorYamlTmpl = `apiVersion: apps/v1
...
 metadata:
   name: {{.ProjectName}}
+  labels:
+    k8s-app: {{.ProjectName}}
 spec:
   replicas: 1
   selector:
     matchLabels:
-      name: {{.ProjectName}}
+      k8s-app: {{.ProjectName}}
```
Comment (fanminshi) [inline pkg/generator/deploy_tmpl.go]: why change `name` to `k8s-app`?

**Comment by fanminshi** [comment]
Could you update the commit message title format as shown in https://github.com/operator-framework/operator-sdk/blob/master/CONTRIBUTING.MD#format-of-the-commit-message?

File: pkg/generator/deploy_tmpl.go (hunk 1 above)
Comment (etiennecoutaud) [reply pkg/generator/deploy_tmpl.go]: I introduced `k8s-app` label for service and servicemonitor. I just changed to have the same label everywhere.

If needed I can change to use `name` label in service and servicemonitor

Comment (fanminshi) [reply pkg/generator/deploy_tmpl.go]: I think we should keep the the original label `name` for now. if change is indeed needed, we can change the label to something else in a future pr.

cc/ @spahl @hasbro17 on label change.

File: pkg/generator/deploy_tmpl.go [hunk 2]
```diff
@@ -33,19 +33,23 @@ const operatorYamlTmpl = `apiVersion: apps/v1
...
     metadata:
       labels:
-        name: {{.ProjectName}}
+        k8s-app: {{.ProjectName}}
     spec:
       containers:
         - name: {{.ProjectName}}
           image: {{.Image}}
+          ports:
+          - containerPort: 9090
```
Comment (eparis) [inline pkg/generator/deploy_tmpl.go]: can we `name` this port? and then use that in the service as `targetPort: name`.  That leaves us flexibility to change the port in the future without having to change the service.

File: pkg/generator/deploy_tmpl.go (hunk 1 above)
Comment (etiennecoutaud) [reply pkg/generator/deploy_tmpl.go]: Okay for using `name` label.
Should I use `name` label in `service` and `servicemonitor` too ?

File: pkg/generator/deploy_tmpl.go (hunk 2 above)
Comment (etiennecoutaud) [reply pkg/generator/deploy_tmpl.go]: Sure ! 👍

**Comment by etiennecoutaud** [comment]
finally change all `k8s-app` labels to `name`

File: pkg/generator/deploy_tmpl.go (hunk 2 above)
Comment (etiennecoutaud) [reply pkg/generator/deploy_tmpl.go]: containerPort is now named `metrics` and use in the service as `targetPort: metrics`

**Comment by eparis** [comment]
This PR LGTM, though it does bother me a little that we have "9090" hard coded in 4 places. Not sure it's worth putting in the template data, but if we really wanted to do a followup.

**Comment by etiennecoutaud** [comment]
I agree with you, an another way could be to add a config file pass to the generator when `operator-sdk new` is running.
We could enable or not prometheus integration, define metrics ports, git init etc ...

**Comment by eparis** [comment]
I'd personally lean pretty hard on 'only give users 1 way to do it and make sure that way is the right way'. I'm definitely in favor of us putting the port into the code and not making it optional. My only (tiny) bit of unease is that it's in the code 4 times instead of 1.

Fewer options make users less able to get into weird situations we don't test or aren't able to help them with.

**Comment by hasbro17** [comment]
We should follow up with the discussion at https://github.com/operator-framework/operator-sdk/issues/222#issuecomment-390311934 before we decide if we want to move forward with this approach.

File: pkg/generator/gen_main_test.go [hunk 3]
```diff
@@ -24,23 +24,46 @@ const mainExp = `package main
...
 	sdk "github.com/operator-framework/operator-sdk/pkg/sdk"
 	k8sutil "github.com/operator-framework/operator-sdk/pkg/util/k8sutil"
 	sdkVersion "github.com/operator-framework/operator-sdk/version"
 
+	"github.com/prometheus/client_golang/prometheus/promhttp"
 	"github.com/sirupsen/logrus"
 )
 
+// Prometheus metrics port
+const promPort = ":9090"
```
Comment (hasbro17) [inline pkg/generator/gen_main_test.go]: Can we use a less common port number at a higher range like 90000.
Also can you define the port number as a global constant somewhere other than the test file like in [k8sutil/constants.go](https://github.com/operator-framework/operator-sdk/blob/master/pkg/util/k8sutil/constants.go). That way we can use it everywhere in the generator package without redefining it.

File: pkg/generator/deploy_tmpl.go (hunk 2 above)
Comment (hasbro17) [reply pkg/generator/deploy_tmpl.go]: Make the port number a variable like `{{.MetricsPort}}` and define a new parameter in [gen_deploy.go](https://github.com/operator-framework/operator-sdk/blob/master/pkg/generator/gen_deploy.go#L60-L63)
```Go
type OperatorYaml struct {
	ProjectName string
	Image       string
	MetricsPort int
}
```
That way we can set MetricsPort to a constant defined elsewhere.

File: pkg/generator/gen_main_test.go (hunk 3 above)
Comment (hasbro17) [reply pkg/generator/gen_main_test.go]: Or something that's actually in the ephemeral port range like 60K. 90K is out of range.

File: pkg/generator/deploy_tmpl.go [hunk 4]
```diff
@@ -54,6 +59,10 @@ spec:
               valueFrom:
                 fieldRef:
                   fieldPath: metadata.namespace
+            - name: OPERATOR_NAME
```
Comment (hasbro17) [inline pkg/generator/deploy_tmpl.go]: Similarly can you make the `OPERATOR_NAME` and `WATCH_NAMESPACE` strings as variables in the template so we can set them using the namespace constants defined in k8sutil/constants.go.
```Go
type OperatorYaml struct {
	ProjectName string
	Image       string
        NameEnv  string
        NamespaceEnv  string
	MetricsPort int
}
```

**Comment by etiennecoutaud** [comment]
@hasbro17 according to your review:
* constant `MetricsPort` is defined in `k8sutil/constant` and set to `60000`
* `OperatorYaml` struct has been updated to embedded environment vars value defined in `k8sutil/constant`
* `Main` struct has also been updated with `MetricsPort` value from constant to use it in template

File: pkg/generator/templates.go [hunk 5]
```diff
@@ -529,14 +551,19 @@ spec:
       containers:
         - name: {{.ProjectName}}
           image: {{.Image}}
+          ports:
+          - containerPort: {{.MetricsPort}}
+            name: metrics
```
Comment (hasbro17) [inline pkg/generator/templates.go]: The string `metrics` is being used here and in `InitOperatorService()`.
We should make this a constant in k8sutil/constants.go.
And then pass it as a parameter `name: {{.MetricsPortName}}`.

File: Gopkg.toml [hunk 6]
```diff
@@ -6,6 +6,10 @@
   name = "github.com/spf13/cobra"
   version = "0.0.2"
 
+[[override]]
+  name = "github.com/prometheus/client_golang"
```
Comment (hasbro17) [inline Gopkg.toml]: This package isn't really used in the SDK itself right now so we can remove this. We'll add it back in later once we actually use this to record default metrics inside the SDK.
For now we can add it to the Gopkg.toml template of the operator:
https://github.com/operator-framework/operator-sdk/blob/master/pkg/generator/templates.go#L372

File: pkg/generator/gen_main_test.go [hunk 7]
```diff
@@ -22,37 +22,57 @@ import (
...
-	"runtime"
+  "context"
+  "runtime"
+  "net/http"
 
-	stub "github.com/example-inc/app-operator/pkg/stub"
-	sdk "github.com/operator-framework/operator-sdk/pkg/sdk"
-	k8sutil "github.com/operator-framework/operator-sdk/pkg/util/k8sutil"
-	sdkVersion "github.com/operator-framework/operator-sdk/version"
+  stub "github.com/example-inc/app-operator/pkg/stub"
```
Comment (hasbro17) [inline pkg/generator/gen_main_test.go]: We should keep the tab size to 4 as it is for all the other templates.

File: pkg/generator/templates.go [hunk 8]
```diff
@@ -130,37 +130,57 @@ spec:
...
-	"runtime"
+  "context"
+  "runtime"
+  "net/http"
 
-	stub "{{.StubImport}}"
-	sdk "{{.OperatorSDKImport}}"
-	k8sutil "{{.K8sutilImport}}"
-	sdkVersion "{{.SDKVersionImport}}"
+  stub "{{.StubImport}}"
```
Comment (hasbro17) [inline pkg/generator/templates.go]: Change tab size back to 4

File: pkg/generator/templates.go [hunk 9]
```diff
@@ -130,37 +130,57 @@ spec:
...
-func main() {
-	printVersion()
+func initOperatorService() {
+  service, err := k8sutil.InitOperatorService()
+  if err != nil {
+    logrus.Fatalf("Failed to init operator service: %v", err)
+  }
+  err = sdk.Create(service)
+  if err != nil {
+    logrus.Infof("Failed to create operator service: %v", err)
```
Comment (hasbro17) [inline pkg/generator/templates.go]: We should ignore the error if the service already exists as it's very likely that the operator pod can restart.
```Go
import "k8s.io/apimachinery/pkg/api/errors"
...
if err != nil && !errors.IsAlreadyExists(err) {
    logrus.Infof("Failed to create operator service: %v", err)
}
```

File: pkg/generator/templates.go [hunk 10]
```diff
@@ -130,37 +130,57 @@ spec:
...
+    return
+  }
+  logrus.Infof("Service %s have been created", service.Name)
```
Comment (hasbro17) [inline pkg/generator/templates.go]:
```Go
logrus.Infof("Metrics service %s created", service.Name)
```

**Comment by hasbro17** [comment]
@etiennecoutaud Thanks for the update. I've tested this out locally and it works fine.
If you can address the above changes we should be good. Also you'll need to rebase since we refactored the generator in #310

**Comment by etiennecoutaud** [comment]
@hasbro17 I completley failed my rebase 😄, anyway the refactor was too heavy. I will close this PR and reopen a clean one.

Is it normally that generator refactor remove all generator unit test ?

**Comment by etiennecoutaud** [comment]
No problem, I close this PR and wait for #315 to be merged.

**Comment by hasbro17** [comment]
@etiennecoutaud Can you reopen this in another PR. I think #315 will only change the main template unit test which should be easy to rebase once it's merged.

Also looking back I think it's best if we keep the generated code to a minimum and keep the port exposing and service creation utility in the SDK.

So the generated main.go should look like:
```Go
func main() {
	printVersion()

	sdk.ExposeMetricsPort()

	resource := "app.example.com/v1alpha1"
	kind := "App"
	namespace, err := k8sutil.GetWatchNamespace()
	if err != nil {
		logrus.Fatalf("Failed to get watch namespace: %v", err)
	}
	resyncPeriod := 5
	logrus.Infof("Watching %s, %s, %s, %d", resource, kind, namespace, resyncPeriod)
	sdk.Watch(resource, kind, namespace, resyncPeriod)
	sdk.Handle(stub.NewHandler())
	sdk.Run(context.TODO())
}
```

And sdk.ExposeMetricsPort() can be defined in `pkg/sdk/metrics.go` as: 
```Go
func ExposeMetricsPort() {
  http.Handle("/"+k8sutil.PrometheusMetricsPortName, promhttp.Handler())
  go http.ListenAndServe(":"+k8sutil. PrometheusMetricsPort, nil)

  service, err := k8sutil.InitOperatorService()
  if err != nil {
    logrus.Fatalf("Failed to init operator service: %v", err)
  }
  err = sdk.Create(service)
  if err != nil && !errors.IsAlreadyExists(err) {
    logrus.Infof("Failed to create operator service: %v", err)
    return
  }
  logrus.Infof("Metrics service %s created", service.Name)
}
```

This way we can actually use the constants for the metrics port and not have them generated as a number and string in the operator's main.go. It also keeps main.go small.
Later on we can change sdk.ExposeMetricsPort() to be configurable by taking the port number as an arg if needed.

**Comment by etiennecoutaud** [comment]
Yes sure, I will before the end of the week.

**Pull request closed without merging**